### PR TITLE
PRJ-194 Support Option+key on Mac clients

### PR DESCRIPTION
**Pull request by SerVB**

### CODE REVIEW

File: projector-agent-ij-injector/src/main/kotlin/org/jetbrains/projector/agent/ijInjector/IjUiUtilsTransformer.kt [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+          {
+            boolean isPrintableMask = true;
+            final java.awt.Toolkit tk = java.awt.Toolkit.getDefaultToolkit();
+            if (tk instanceof sun.awt.SunToolkit) {
+              final int mod = $1.getModifiers();
+              isPrintableMask = ((sun.awt.SunToolkit) tk).isPrintableCharacterModifiersMask(mod);
+            }
+
+            final char c = $1.getKeyChar();
+            return (isPrintableMask && (c >= 0x20) && (c != 0x7F)) || (!isPrintableMask && (c >= 0x200C) && (c <= 0x200D));
```
Comment (ARTI1208) [inline projector-agent-ij-injector/src/main/kotlin/org/jetbrains/projector/agent/ijInjector/IjUiUtilsTransformer.kt:81]: I think comments from superclass method on these constants could be helpful for future readers

File: projector-agent-ij-injector/src/main/kotlin/org/jetbrains/projector/agent/ijInjector/IjUiUtilsTransformer.kt [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    ) {
+      logger.debug { "agentmain start" }
+
+      val extensionPointName = utils.createExtensionPointName(SE_CONTRIBUTOR_EP)
+      val extensions = utils.extensionPointNameGetExtensions(extensionPointName)
+
+      val ideClassloader = extensions.filterNotNull().first()::class.java.classLoader
+
+      val ideCp = ClassPool().apply {
+        appendClassPath(LoaderClassPath(ideClassloader))
```
Comment (ARTI1208) [inline projector-agent-ij-injector/src/main/kotlin/org/jetbrains/projector/agent/ijInjector/IjUiUtilsTransformer.kt:109]: You can pass the classloader of current class as ProjectorClassLoader knows about IDE classes

File: projector-agent-ij-injector/src/main/kotlin/org/jetbrains/projector/agent/ijInjector/IjUiUtilsTransformer.kt [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    return clazz.toBytecode()
+  }
+
+  companion object {
+
+    private val logger = Logger<IjUiUtilsTransformer>()
+
+    private const val SE_CONTRIBUTOR_EP = "com.intellij.searchEverywhereContributor"
+
+    private const val uiUtilClass = "com.intellij.util.ui.UIUtil"
```
Comment (ARTI1208) [inline projector-agent-ij-injector/src/main/kotlin/org/jetbrains/projector/agent/ijInjector/IjUiUtilsTransformer.kt:95]: You can directly reference this class now, just add `com.jetbrains.intellij.platform:util-ui` compileOnly dependency (or `com.jetbrains.intellij.platform:ide-impl` to get less warnings viewing UIUtil sources)

**Review by ARTI1208 — CHANGES_REQUESTED** [review]
Also I don't have any macOS device and can't check it actually works (or is there a way to emulate macOS behavior without VM?), but I assume you tested this

File: projector-agent-ij-injector/src/main/kotlin/org/jetbrains/projector/agent/ijInjector/IjUiUtilsTransformer.kt (hunk 1 above)
Comment (SerVB) [reply projector-agent-ij-injector/src/main/kotlin/org/jetbrains/projector/agent/ijInjector/IjUiUtilsTransformer.kt:81]: There are no comments in the superclass actually: [DefaultEditorKit.java#L873-L884](https://github.com/JetBrains/JetBrainsRuntime/blob/04f1cf9b478ad555a0a6062da25705eb0fa0769f/src/java.desktop/share/classes/javax/swing/text/DefaultEditorKit.java#L873-L884). I guess a reference `like in javax.swing.text.DefaultEditorKit.DefaultKeyTypedAction` is enough here. Unless we talk about different things. Could you please clarify?

Comment (ARTI1208) [reply projector-agent-ij-injector/src/main/kotlin/org/jetbrains/projector/agent/ijInjector/IjUiUtilsTransformer.kt:81]: Oh, missed it, `class UIUtil { public static boolean isReallyTypedEvent` 3 lines down mislead me.. It's ok then

**Comment by SerVB** [comment]
> is there a way to emulate macOS behavior without VM?

¯\_(ツ)_/¯

> I assume you tested this

Yep, I've tested and it's OK if you take a look only at code. Also, you can check that nothing is broken on Linux, if you want.

> You can pass the classloader of current class as ProjectorClassLoader knows about IDE classes
> You can directly reference this class now

Actually, I've copied the code of this class from another transformer in the same package. I understood that it's not OK to have copy-pasted code, so I was going to ask you to refactor transformers after this PR ;) Not only new classloaders can be used, but also duplicated code can be removed and logging could be implemented better (just like a list of successful and unsuccessful transformations in the console).

I guess it's OK to leave this class like this for now until the refactoring. Do you think I should clean up this class now anyway?

File: projector-agent-ij-injector/src/main/kotlin/org/jetbrains/projector/agent/ijInjector/IjUiUtilsTransformer.kt (hunk 1 above)
Comment (SerVB) [reply projector-agent-ij-injector/src/main/kotlin/org/jetbrains/projector/agent/ijInjector/IjUiUtilsTransformer.kt:81]: No problem. By the way, I have some problem with terminology here: `DefaultEditorKit.DefaultKeyTypedAction` isn't a superclass, it's just a default implementation class 😅

**Comment by ARTI1208** [comment]
Seems reasonable to do it for all transformers at one time

**Review by ARTI1208 — APPROVED** [review]

**Pull request merged**